docs/frontend-performance.md

```diff
@@ -0,0 +1,2984 @@
+      ))}
+    </div>
+  );
+};
+```
+
+**Key rules:**
+
+- ✅ Use unique IDs from data (address, id, uuid)
+- ✅ Keys must be stable across re-renders
+- ⚠️ Only use index if list never reorders and items don't have IDs
+- ❌ Never use random values or `Math.random()`
+
+### ❌ Anti-Pattern: Using Index as Key for Dynamic Lists
+
+Using array index as key breaks React's reconciliation when lists can be reordered, filtered, or items added/removed.
+
+```typescript
+❌ WRONG: Index keys break reconciliation
+const TokenList = ({ tokens }: TokenListProps) => {
+  // If tokens can be reordered/filtered, this breaks React's reconciliation
+  return (
+    <div>
+      {tokens.map((token, index) => (
+        <TokenItem key={index} token={token} /> // Bad!
+      ))}
+    </div>
+  );
+};
+
+✅ CORRECT: Use unique, stable identifiers
+const TokenList = ({ tokens }: TokenListProps) => {
+  return (
+    <div>
+      {tokens.map(token => (
+        <TokenItem key={token.address} token={token} /> // Good!
+      ))}
+    </div>
+  );
+};
+```
+
+**Problems with index keys:**
+
+- React can't track items when list reorders
+- State gets mixed up between items
+- Causes bugs with form inputs, focus, animations
+- Performance issues from unnecessary re-renders
+
+**Solution:** Always use unique, stable identifiers from your data (address, id, uuid). Only use index if the list is static and never reorders.
+
+### Virtualize Long Lists
+
+For lists with 100+ items, use virtualization to only render visible items.
+
+```typescript
+❌ WRONG: Rendering 1000+ items at once
+const TransactionList = ({ transactions }: TransactionListProps) => {
+  return (
+    <div className="transaction-list">
+      {transactions.map(tx => (
+        <TransactionItem key={tx.hash} transaction={tx} />
+      ))}
+    </div>
+  );
+};
+
+✅ CORRECT: Use virtualization library
+import { FixedSizeList } from 'react-window';
+
+const TransactionList = ({ transactions }: TransactionListProps) => {
+  const Row = ({ index, style }: { index: number; style: React.CSSProperties }) => (
+    <div style={style}>
+      <TransactionItem transaction={transactions[index]} />
+    </div>
+  );
+
+  return (
+    <FixedSizeList
+      height={600}
+      itemCount={transactions.length}
+      itemSize={80}
+      width="100%"
+    >
+      {Row}
+    </FixedSizeList>
+  );
+};
+```
+
+```typescript
+✅ GOOD: Virtual scrolling for 1000+ assets
+import { FixedSizeList } from 'react-window';
+
+const AssetList = ({ assets }: AssetListProps) => {
+  const Row = ({ index, style }: { index: number; style: React.CSSProperties }) => (
+    <div style={style}>
+      <Asset asset={assets[index]} />
+    </div>
+  );
+
+  return (
+    <FixedSizeList
+      height={600}
+      itemCount={assets.length}
+      itemSize={80}
+      width="100%"
+    >
+      {Row}
+    </FixedSizeList>
+  );
+};
+```
+
+**Recommended libraries:**
+
+- `react-window` - Lightweight, recommended for most use cases
```

Comment:
I would update this to `@tanstack/react-virtual`, the one we recently started using for the Tokens list.

- It's underlying library is framework agnostic (`@tanstack/virtual`) 
- both `react-window` and `react-virtualized` are framework specific, and the latter hasn't had significant updates in years